src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1075,8 +1079,33 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
             }
 
             diagnostics.AddRangeAndFree(localDiagnostics);
-            var boundDeclType = new BoundTypeExpression(typeSyntax, aliasOpt, inferredType: isVar, type: declTypeOpt.TypeSymbol);
-            return new BoundLocalDeclaration(associatedSyntaxNode, localSymbol, boundDeclType, initializerOpt, arguments, hasErrors);
+
+            BoundTypeExpression boundDeclType = null;
+
+            if (includeBoundType)
+            {
+                var invalidDimensions = ArrayBuilder<BoundExpression>.GetInstance();
+
+                typeSyntax.VisitTypeSyntaxes(type =>
+                {
+                    if (type is ArrayTypeSyntax arrayTypeSyntax)
+                    {
+                        foreach (var rankSpecifier in arrayTypeSyntax.RankSpecifiers)
+                        foreach (var expressionSyntax in rankSpecifier.Sizes)
+                        {
+                            var size = BindArrayDimension(expressionSyntax, typeSyntax, diagnostics);
```

Comment:
>typeSyntax, diagnostics [](start = 76, length = 23)

Instead of relying on capturing locals in a closure, consider passing all values that a need for the lambda as an explicit parameter (perhaps a tuple). This will avoid closure allocation and will allow the delegate instance to be cached and reused.

Reply:
For example, see the following call in Binder_Expressions.cs:
```
boundInitExprOpt = boundInitExprOpt.SelectAsArray((expr, t) => GenerateConversionForAssignment(t.elementType, expr, t.diagnostics), (elementType, diagnostics));
```

---
In reply to: [260051709](https://github.com/dotnet/roslyn/pull/32843#discussion_r260051709) [](ancestors = 260051709)

Reply:
There are a total of 4 arguments, which is a quite a large tuple to pass around a lot.
Do you think it might be better to create a threadlocal Tuple to pass around the arguments, rather than using a ValueTuple?

Reply:
please don't use threadlocals.  The pattern aleksey mentioned is an appropriate way to do things.

Reply:
>There are a total of 4 arguments, which is a quite a large ValueTuple to pass around a lot.

I think this is not likely to make any noticeable impact because deep nesting of the type syntax won't be common.  


---
In reply to: [260198786](https://github.com/dotnet/roslyn/pull/32843#discussion_r260198786) [](ancestors = 260198786)